adding validations on kubelet starting configurations

huangjiuyuan opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
-->

**What this PR does / why we need it**:
I found some validations of kubelet starting options were missing when I was creating a custom cluster from scratch. The kubelet does not check invalid configurations on `--cadvisor-port`, `--event-burst`, `--image-gc-high-threshold`, etc. I have added some validations in kubelet like validations in `cmd/kube-apiserver/app/options/validation.go`.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
Adds additional validation for kubelet in `pkg/kubelet/apis/kubeletconfig/validation`.
```

Commit:
adding validations on kubelet starting configurations

xiangpengzhao commented on cmd/kubelet/app/server.go:
```diff
@@ -316,13 +317,44 @@ func initConfigz(kc *componentconfig.KubeletConfiguration) (*configz.Config, err
-		return fmt.Errorf("invalid configuration: system container was specified and cgroup root was not specified")
+		errors = append(errors, fmt.Errorf("Invalid configuration: system container was specified and cgroup root was not specified"))
+	}
+	if s.CAdvisorPort < 0 || s.CAdvisorPort > 65535 {
```
You may want to use `IsValidPortNum` (k8s.io/apimachinery/pkg/util/validation) instead for these port validations.

allencloud commented:
I think we had better to validate config of component kubelet in `pkg/api/componentconfig`.

xiangpengzhao commented on cmd/kubelet/app/server.go:
```diff
@@ -316,13 +317,44 @@ func initConfigz(kc *componentconfig.KubeletConfiguration) (*configz.Config, err
+		errors = append(errors, fmt.Errorf("Invalid configuration: system container was specified and cgroup root was not specified"))
+	}
+	if s.CAdvisorPort < 0 || s.CAdvisorPort > 65535 {
+		errors = append(errors, fmt.Errorf("Invalid configuration: --cadvisor-port %v must be between 0 and 65535", s.CAdvisorPort))
```
s/`between 0 and 65535`/`between 1 and 65535`

xiangpengzhao replied:
also other ports validations below.

xiangpengzhao reviewed (changes requested):
add unit tests for the validation?

xiangpengzhao commented:
>I think we had better to validate config of component kubelet in pkg/api/componentconfig.

Agreed. This should wait for https://github.com/kubernetes/kubernetes/pull/49087 merged.

xiangpengzhao commented on cmd/kubelet/app/server.go:
```diff
@@ -316,25 +318,81 @@ func initConfigz(kc *componentconfig.KubeletConfiguration) (*configz.Config, err
+		errors = append(errors, fmt.Errorf("Invalid configuration: system container was specified and cgroup root was not specified"))
 	}
+	if utilvalidation.IsValidPortNum(int(s.CAdvisorPort)) != nil {
+		errors = append(errors, fmt.Errorf("Invalid configuration: --cadvisor-port %v must be between 0 and 65535", s.CAdvisorPort))
```
this is still not fixed. `between 0 and 65535` should be `between 1 and 65535`

huangjiuyuan commented:
Hi, @xiangpengzhao 
I have fixed the port range errors and added an unit test for validations. If https://github.com/kubernetes/kubernetes/pull/49087 gets merged, I will move all the validations to `pkg/apis/componentconfig`.

xiangpengzhao commented:
@huangjiuyuan SGTM :)

mtaufen commented:
@huangjiuyuan thanks a lot for this work! 

Please wait for https://github.com/kubernetes/kubernetes/pull/49051 to merge, then move your validation to `pkg/kubelet/apis/kubeletconfig/validation`.

mtaufen commented:
/ok-to-test

Probably also a good idea to add a release note for the fact that the Kubelet has additional validation.

huangjiuyuan commented:
Hi, @mtaufen 
Thank you for your wonderful work!

mtaufen commented:
@huangjiuyuan are you still working on this? Can you move the validation to pkg/kubelet/apis/kubeletconfig/validation?

huangjiuyuan commented:
Hi @mtaufen 
Yes, I am still working on this one. I will move the validations to `pkg/kubelet/apis/kubeletconfig/validation` today.

huangjiuyuan commented:
Hi @mtaufen
Could you please take a look at this?

mtaufen commented:
I will review this Monday afternoon. Thanks for being patient :).

mtaufen commented on pkg/kubelet/apis/kubeletconfig/validation/validation.go:
```diff
@@ -41,6 +42,60 @@ func ValidateKubeletConfiguration(kc *kubeletconfig.KubeletConfiguration) error
 		return fmt.Errorf("invalid configuration: system container was specified and cgroup root was not specified")
 	}
+	if utilvalidation.IsValidPortNum(int(kc.CAdvisorPort)) != nil {
+		return fmt.Errorf("Invalid configuration: --cadvisor-port %v must be between 1 and 65535", kc.CAdvisorPort)
```
Please include the field name and the flag name in these error messages. The configuration can come from a set of flags or a file containing a serialized KubeletConfiguration. Eventually the flags will be removed, and then we'll remove the flag part of these error messages too. Also, please always specify inclusive or exclusive when describing ranges.

For now, this should do: `Invalid configuration: CAdvisorPort (--cadvisor-port) %v must be between 1 and 65535, inclusive`. And similarly for all the other errors.

mtaufen commented on pkg/kubelet/apis/kubeletconfig/validation/validation.go:
```diff
@@ -41,6 +42,60 @@ func ValidateKubeletConfiguration(kc *kubeletconfig.KubeletConfiguration) error
+		return fmt.Errorf("Invalid configuration: --cadvisor-port %v must be between 1 and 65535", kc.CAdvisorPort)
+	}
+	if kc.EventBurst < 0 {
+		return fmt.Errorf("Invalid configuration: --event-burst %v should not be a negative number", kc.EventBurst)
```
nit: s/should/must
see [rfc2119](https://www.ietf.org/rfc/rfc2119.txt)
and similarly for other instances where "should" actually means "must," according to this validation code.

mtaufen commented on pkg/kubelet/apis/kubeletconfig/validation/validation_test.go:
```diff
@@ -0,0 +1,81 @@
+		RegistryBurst:               -10,
+		RegistryPullQPS:             -10,
+	}
+	if err := ValidateKubeletConfiguration(errorCase); err == nil {
```
I'd prefer we test that all of the validation works correctly, rather than that we get at least one error.
It would be better to aggregate the errors from validation and return all of them, then you could check for every expected error as part of your test.

mtaufen replied:
`k8s.io/apimachinery/pkg/util/validation/field` has an `ErrorList` type that you can use for this, see for example what @allencloud did in #49087.

mtaufen commented on staging/src/k8s.io/apimachinery/pkg/util/validation/validation_test.go:
```diff
@@ -154,6 +154,22 @@ func TestIsValidPortNum(t *testing.T) {
 }
 
+func TestIsInRange(t *testing.T) {
+	goodValues := [][]int{{1, 0, 10}, {5, 5, 20}, {25, 10, 25}}
```
suggest:
```
goodValues := []struct{
  value int
  min int
  max int
}{{1, 0, 10}, {5, 5, 20}, {25, 10, 25}}
```
and then you can do
```
IsInRange(val.value, val.min. val.max)
```
which reads more clearly.
Similarly for `badValues`.

mtaufen commented on staging/src/k8s.io/apimachinery/pkg/util/validation/validation_test.go:
```diff
@@ -154,6 +154,22 @@ func TestIsValidPortNum(t *testing.T) {
+	goodValues := [][]int{{1, 0, 10}, {5, 5, 20}, {25, 10, 25}}
+	for _, val := range goodValues {
+		if msgs := IsInRange(val[0], val[1], val[2]); len(msgs) != 0 {
+			t.Errorf("expected true for %d, got %v", val, msgs)
```
`IsInRange` doesn't actually return a boolean, I think you expect `nil` here.
Similarly for `badValues` but with the reverse expectation.

mtaufen reviewed (changes requested):
I had a few comments. Thanks for this work :smile:.

huangjiuyuan commented:
/test pull-kubernetes-e2e-gce-etcd3
/test pull-kubernetes-e2e-kops-aws

huangjiuyuan commented:
/test pull-kubernetes-e2e-kops-aws

huangjiuyuan commented:
@mtaufen Thanks for your wonderful suggestions. I have used an aggregator to return all errors when validating. Could you please take a quick look?

huangjiuyuan commented:
/test pull-kubernetes-e2e-kops-aws

huangjiuyuan commented:
/test pull-kubernetes-e2e-kops-aws

yujuhong commented:
/approve

huangjiuyuan commented:
/test pull-kubernetes-e2e-kops-aws

mtaufen commented on staging/src/k8s.io/apimachinery/pkg/util/validation/validation_test.go:
```diff
@@ -154,6 +154,30 @@ func TestIsValidPortNum(t *testing.T) {
+		max   int
+	}{{1, 0, 10}, {5, 5, 20}, {25, 10, 25}}
+	for _, val := range goodValues {
+		if msgs := IsInRange(val.value, val.min, val.max); len(msgs) != 0 {
```
nit: `len(msgs) > 0` reads better

mtaufen commented on staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go:
```diff
@@ -188,6 +188,14 @@ func IsValidPortNum(port int) []string {
 	return []string{InclusiveRangeError(1, 65535)}
 }
 
+// IsInRange tests that the argument is in a range inclusively.
```
s/a range inclusively/an inclusive range

mtaufen commented on staging/src/k8s.io/apimachinery/pkg/util/validation/validation_test.go:
```diff
@@ -154,6 +154,30 @@ func TestIsValidPortNum(t *testing.T) {
+	}{{1, 0, 10}, {5, 5, 20}, {25, 10, 25}}
+	for _, val := range goodValues {
+		if msgs := IsInRange(val.value, val.min, val.max); len(msgs) != 0 {
+			t.Errorf("expected true for %d, got %v", val, msgs)
```
again, I don't like the use of the boolean words `true` and `false` in these errors, because IsInRange doesn't deal in booleans
suggest: `t.Errorf("expected no errors for %#v, but got %v", val, msgs)`
`%d` isn't correct because val is not an integer.
similarly for `badValues`

mtaufen reviewed (changes requested):
I had some additional comments. Thanks for switching to the aggregated error model.

huangjiuyuan commented:
/test pull-kubernetes-unit

huangjiuyuan commented:
@mtaufen Thank you for the patience. Could you take a look?

mtaufen commented:
/lgtm

mtaufen commented:
/assign @liggitt 
for apimachinery approval

mtaufen commented:
/approve no-issue

liggitt commented on pkg/kubelet/apis/kubeletconfig/validation/validation.go:
```diff
@@ -19,17 +19,75 @@ package validation
-		return fmt.Errorf("invalid configuration: system container was specified and cgroup root was not specified")
+		allErrors = append(allErrors, fmt.Errorf("Invalid configuration: SystemCgroups (--system-cgroups) was specified and CgroupRoot (--cgroup-root) was not specified"))
+	}
+	if utilvalidation.IsValidPortNum(int(kc.CAdvisorPort)) != nil {
```
what are you supposed to set the flag to to disable cadvisor? I thought 0

mtaufen replied:
Good catch @liggitt!

liggitt commented on pkg/kubelet/apis/kubeletconfig/validation/validation.go:
```diff
@@ -19,17 +19,75 @@ package validation
+	if utilvalidation.IsValidPortNum(int(kc.Port)) != nil {
+		allErrors = append(allErrors, fmt.Errorf("Invalid configuration: Port (--port) %v must be between 1 and 65535, inclusive", kc.Port))
+	}
+	if utilvalidation.IsValidPortNum(int(kc.ReadOnlyPort)) != nil {
```
same here, should be able to disable

mtaufen commented:
/lgtm cancel
 - see @liggit's comments

huangjiuyuan commented:
@liggitt Thanks for your wonderful suggestions. The port validations has been updated.

liggitt commented:
/approve no-issue

apimachinery changes LGTM

@mtaufen please double check the ranges to make sure we're not invalidating use of 0 or -1 that currently means "use default" or "disable", etc.

mtaufen commented on pkg/kubelet/apis/kubeletconfig/validation/validation.go:
```diff
@@ -19,17 +19,75 @@ package validation
+	if kc.EventRecordQPS < 0 {
+		allErrors = append(allErrors, fmt.Errorf("Invalid configuration: EventRecordQPS (--event-qps) %v must not be a negative number", kc.EventRecordQPS))
+	}
+	if utilvalidation.IsValidPortNum(int(kc.HealthzPort)) != nil {
```
This should also be possible to disable by passing 0.

mtaufen commented on pkg/kubelet/apis/kubeletconfig/validation/validation.go:
```diff
@@ -19,17 +19,75 @@ package validation
-		return fmt.Errorf("invalid configuration: system container was specified and cgroup root was not specified")
+		allErrors = append(allErrors, fmt.Errorf("Invalid configuration: SystemCgroups (--system-cgroups) was specified and CgroupRoot (--cgroup-root) was not specified"))
+	}
+	if kc.CAdvisorPort < 0 || kc.CAdvisorPort > 65535 {
```
You can still use the utility that defines the port ranges
`if kc.CAdvisorPort != 0 && utilvalidation.IsValidPortNum(int(kc.CAdvisorPort)) != nil`

mtaufen commented on pkg/kubelet/apis/kubeletconfig/validation/validation.go:
```diff
@@ -19,17 +19,75 @@ package validation
+	if kc.PodsPerCore < 0 {
+		allErrors = append(allErrors, fmt.Errorf("Invalid configuration: PodsPerCore (--pods-per-core) %v must not be a negative number", kc.PodsPerCore))
+	}
+	if kc.Port < 0 || kc.Port > 65535 {
```
I think kc.Port has to be a true valid port range, and you disable the server by setting `kc.EnableServer = false`

huangjiuyuan commented:
/test pull-kubernetes-e2e-gce-bazel

huangjiuyuan commented:
@mtaufen Good to go?

mtaufen commented:
/lgtm

mtaufen commented:
/retest

huangjiuyuan commented:
/test pull-kubernetes-federation-e2e-gce

huangjiuyuan commented:
/test pull-kubernetes-federation-e2e-gce

mtaufen commented:
/retest
